msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java

```diff
@@ -1276,8 +1276,16 @@ private void initializeApplication() throws MsalClientException {
         final Context context = mPublicClientConfiguration.getAppContext();
         setupTelemetry(context, mPublicClientConfiguration);
 
-        AzureActiveDirectory.setEnvironment(mPublicClientConfiguration.getEnvironment());
-        Authority.addKnownAuthorities(mPublicClientConfiguration.getAuthorities());
+        // Today, this method runs on the main thread. AzureActiveDirectory.setEnvironment() and Authority.addKnownAuthorities()
```

Comment:
These 2 values should actually be made as part of parameter list, not a global variable.

Putting them in the background like this could also result in a race condition. (the "read" part happens before this "write" part is done) 

Reply:
+1 on the race condition.

That said moving them to parameter list is likely a big change.

Reply:
Could we perhaps use a future here to block on the completion of the thread. Do we know if that would also cause ANRs?

Reply:
> +1 on the race condition.
> 
> That said moving them to parameter list is likely a big change.

Yeah, I spent some time yesterday trying to move it away from being a global variable, it's a very big change. Doing that will require a concentrated refactoring effort, especially if we are also going to do it for knownAuthorities while we are at it.

Reply:
> Could we perhaps use a future here to block on the completion of the thread. Do we know if that would also cause ANRs?

Anything that block the main thread is capable of casuing an ANR. The idea of this PR is to remove all locks from the main thread during the PCA init to remove the possibility of an ANR, but it introduces a possibility of a race condition

Reply:
I'm currently working on another approach with the help Dome, where we separate the locks used in AzureActiveDirectory, as the main cause of the ANR in this case is the lock used to fetch/set environment, is the exact same lock that is used for the performCloudDiscovery network call. If we can remove that pairing, the ANR should become impossible

Reply:
For the sake of the ANR, we can also consider just moving the callback handling to the background, so that all of the PCA intialization happens on the same thread, but nothing is locked on main. We aren't directly returning anything here just communicating to the listener. I don't think this will create any race conditions. We aren't directly returning anything here just communicating to the listener. thoughts @rpdome @shahzaibj